redshift/redshift.go

```diff
@@ -19,59 +17,55 @@ type dbExecCloser interface {
 	Exec(query string, args ...interface{}) (sql.Result, error)
 }
 
+// S3Info holds the information necessary to copy data from s3 buckets
+type S3Info struct {
+	Region    string
+	AccessID  string
+	SecretKey string
+}
+
 // Redshift wraps a dbExecCloser and can be used to perform operations on a redshift database.
 type Redshift struct {
 	dbExecCloser
-	accessID  string
-	secretKey string
+	s3Info S3Info
 }
 
 var (
-	// TODO: include flag validation
-	awsAccessKeyID     = env.MustGet("AWS_ACCESS_KEY_ID")
-	awsSecretAccessKey = env.MustGet("AWS_SECRET_ACCESS_KEY")
-	host               = flag.String("redshifthost", "", "Address of the redshift host")
-	port               = flag.Int("redshiftport", 0, "Address of the redshift host")
-	db                 = flag.String("redshiftdatabase", "", "Redshift database to connect to")
-	user               = flag.String("redshiftuser", "", "Redshift user to connect as")
-	pwd                = flag.String("redshiftpassword", "", "Password for the redshift user")
-	timeout            = flag.Duration("redshiftconnecttimeout", 10*time.Second,
-		"Timeout while connecting to Redshift. Defaults to 10 seconds.")
 )
 
-// NewRedshift returns a pointer to a new redshift object using configuration values set in the
-// flags.
-func NewRedshift() (*Redshift, error) {
+// NewRedshift returns a pointer to a new redshift object using configuration values passed in
+// on instantiation and the AWS env vars we assume exist
+// Don't need to pass s3 info unless doing a COPY operation
+func NewRedshift(host, port, db, user, pwd string, timeout int, s3Info S3Info) (*Redshift, error) {
 	flag.Parse()
-	source := fmt.Sprintf("host=%s port=%d dbname=%s connect_timeout=%d", *host, *port, *db, int(timeout.Seconds()))
-	log.Println("Connecting to Reshift Source: ", source)
-	source += fmt.Sprintf(" user=%s password=%s", *user, *pwd)
+	source := fmt.Sprintf("host=%s port=%d dbname=%s connect_timeout=%d", host, port, db, timeout)
+	log.Println("Connecting to Redshift Source: ", source)
+	source += fmt.Sprintf(" user=%s password=%s", user, pwd)
 	sqldb, err := sql.Open("postgres", source)
 	if err != nil {
 		return nil, err
 	}
-	return &Redshift{sqldb, awsAccessKeyID, awsSecretAccessKey}, nil
+	return &Redshift{sqldb, s3Info}, nil
 }
 
-func (r *Redshift) logAndExec(cmd string, creds bool) (sql.Result, error) {
+func (r *Redshift) logAndExec(cmd string) (sql.Result, error) {
 	log.Print("Executing Redshift command: ", cmd)
-	if creds {
-		cmd += fmt.Sprintf(" CREDENTIALS '%s=%s;%s=%s'", "aws_access_key_id", r.accessID, "aws_secret_access_key", r.secretKey)
-	}
```

Comment:
The logs never printed out credentials. The logging happens before the conditional. This affects the actual command that needs to be run.